### ISSUE TITLE
Flaky test: SuspendedTabletsIT

### ISSUE THREAD

**Issue by ctubbsii** [apache/accumulo#1791]
**Test name(s)**
 - org.apache.accumulo.test.master.SuspendedTabletsIT.shutdownAndResumeTserver

**Describe the failure observed**
The test failed with an assertion error on the first run, and passed on the second. It should pass every time:
```java
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 318.472 s <<< FAILURE! - in org.apache.accumulo.test.master.SuspendedTabletsIT
[ERROR] shutdownAndResumeTserver(org.apache.accumulo.test.master.SuspendedTabletsIT)  Time elapsed: 50.098 s  <<< FAILURE!
java.lang.AssertionError: expected:<[1;18;17, 1;1<, 1;10;1, 1;23;22, 1;24;23, 1;25;24, 1;26;25, 1;27;26, 1;17;16, 1;28;27]> but was:<[1;23;22, 1;27;26, 1;24;23, 1;28;27]>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.apache.accumulo.test.master.SuspendedTabletsIT.suspensionTestBody(SuspendedTabletsIT.java:243)
	at org.apache.accumulo.test.master.SuspendedTabletsIT.shutdownAndResumeTserver(SuspendedTabletsIT.java:117)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

**Testing Environment:**
 - Version of this project: 2.1.0-SNAPSHOT
 - First commit known to fail (or current commit): 5cb30ed0e9a7f3d481ea49dd126d64cc7ab9ef53
 - Jenkins shows:
```
Executing Maven:  -B -f /home/jenkins/workspace/Accumulo/main/pom.xml -V -e -ntp clean verify -Dstyle.color=always -Dfailsafe.rerunFailingTestsCount=5 -Dsurefire.forkCount=2C
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/asfpackages/maven/apache-maven-3.6.3
Java version: 11.0.6, vendor: Oracle Corporation, runtime: /usr/local/asfpackages/java/jdk-11.0.6
Default locale: en_US, platform encoding: ISO-8859-1
OS name: "linux", version: "4.15.0-74-generic", arch: "amd64", family: "unix"
```

**Comment by ctubbsii** [comment]
@milleruntime  I assigned this issue to you, because you self-assigned #1806, which I closed as a duplicate of this one. The impetus for creating these two issues may have been separate failures... but there's no reason to track its failures on two separate issues.

**Comment by DomGarguilo** [comment]
This test fails for me more often then it passes. I got the same error as you for shutDownAndResumeTserver once but most of the time this test reaches the 5 minute timeout for me. It did pass once in which case the crashAndResumeTserver test failed with similar error: 
```java
java.lang.AssertionError: 
Expected :[1;11;10]
Actual   :[1;18;17, 1;11;10, 1;19;18, 1;12;11, 1;13;12, 1;14;13, 1;15;14, 1;16;15, 1;17;16, 1;8;7]

	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.apache.accumulo.test.master.SuspendedTabletsIT.suspensionTestBody(SuspendedTabletsIT.java:225)
	at org.apache.accumulo.test.master.SuspendedTabletsIT.crashAndResumeTserver(SuspendedTabletsIT.java:100)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
It looks like in both cases an assertEquals() of deadTaletServer and another server is failing. In both cases either the expected value is a subset of the actual value or vise-versa. I'm not too sure what this means yet but im looking into it.

**Comment by DomGarguilo** [comment]
@milleruntime Are you working on this? I noticed you are assigned to this ticket.

**Comment by milleruntime** [comment]
> @milleruntime Are you working on this? I noticed you are assigned to this ticket.

Not at the moment, no, knock yourself out!  I had looked into it some last year but don't remember if I got anywhere.  My guess is that since I don't have a branch for it, I did not find anything significant.   I assigned it to you

**Comment by DomGarguilo** [comment]
The `assertEquals` that is failing compares the equivalence of the tablets on a server before and after that server is suspended. There is a comment before the assert that reads: 
> By this point, all tablets should be either hosted or suspended. All suspended tablets should "belong" to the dead tablet servers, and should be in exactly the same place as before any tserver death.

I'm not too sure if this statement is true however I have observed tablets travelling between servers before and after being suspended. Below is a printout of the list of tablets on two servers before and after suspension:
```java
2021-01-19T14:27:28,658 [master.SuspendedTabletsIT] DEBUG: Suspended thor:42727: [1;29;28, 1;1<, 1;10;1, 1;11;10, 1;12;11, 1;24;23, 1;4;3, 1;15;14, 1;3;29, 1;27;26, 1;28;27, 1;8;7]
2021-01-19T14:27:28,664 [master.SuspendedTabletsIT] DEBUG: Previously hosted thor:42727: [1;29;28, 1;1<, 1;10;1, 1;11;10, 1;12;11, 1;15;14, 1;3;29, 1;27;26, 1;28;27, 1;8;7]
2021-01-19T14:27:28,664 [master.SuspendedTabletsIT] DEBUG: Suspended thor:44931: [1;21;20, 1;20;2, 1;22;21, 1;23;22, 1;25;24, 1;5;4, 1;26;25, 1;16;15]
2021-01-19T14:27:28,664 [master.SuspendedTabletsIT] DEBUG: Previously hosted thor:44931: [1;21;20, 1;20;2, 1;22;21, 1;23;22, 1;24;23, 1;4;3, 1;25;24, 1;5;4, 1;26;25, 1;16;15]
```
Its a bit hard to read but some tablets that appeared on server A are appearing on server B and vise-versa. This tells me that either the assert in this test is making an incorrect assumption about the nature of tablets on a suspended server, or the test is behaving correctly and catching a valid error. Any thoughts would be helpful.

**Comment by ctubbsii** [comment]
> This tells me that either the assert in this test is making an incorrect assumption about the nature of tablets on a suspended server, or the test is behaving correctly and catching a valid error. Any thoughts would be helpful.

That is a correct interpretation of the reported error. The intent of the suspending feature is to ensure that tablets get reassigned to the same server as before, if the outage is brief enough. One would need to investigate to determine why they are migrating... but here are a few possible scenarios:

* the suspend time elapsed and the tablets are free to migrate before we are able to check (possibly because the tserver took too long to restart or WAL recovery on tablet load is causing the suspend time to elapse before we check the tablet states)
* the tserver fully recovered and then participated in subsequent migrations
* we're splitting tablets and creating new migrations (possibly because splits weren't stabilized before killing the tserver)
* the balancer is misbehaving, and rebalancing when it isn't supposed to

**Comment by DomGarguilo** [comment]
The issue may be caused by this section of code:
```java
      // Wait for all of the tablets to hosted ...
      log.info("Waiting on hosting and balance");
      TabletLocations ds;
      for (ds = TabletLocations.retrieve(ctx, tableName); ds.hostedCount != TABLETS;
          ds = TabletLocations.retrieve(ctx, tableName)) {
        Thread.sleep(1000);
      }

      // ... and balanced.
      ctx.instanceOperations().waitForBalance();
      do {
        // Give at least another 5 seconds for migrations to finish up
        Thread.sleep(5000);
        ds = TabletLocations.retrieve(ctx, tableName);
      } while (ds.hostedCount != TABLETS);

      // Pray all of our tservers have at least 1 tablet.
      assertEquals(TSERVERS, ds.hosted.keySet().size());

      // Kill two tablet servers hosting our tablets. This should put tablets into suspended state,
      // and thus halt balancing.

      TabletLocations beforeDeathState = ds;

      serverStopper.eliminateTabletServers(ctx, beforeDeathState, 2);
```
The problem may be coming from the amount of time that is waited for migrations to finish. In the for loop, the break condition is the same as the do-while loop meaning once the for loop is finished the do-while will only execute once(essentially rendering it useless). This may be an issue since the migrations are only ever given 5 seconds to complete before the pre-suspension state of the tablets are recorded. After this happens, 2 servers are killed which sometimes takes a bit of time. This may be the window of time in which the tablets migrate under the radar.

I have tried increasing this amount of time and haven't gotten the error since, however this error doesn't happen too often as is so its hard to tell if this is definitely the cause.

**Comment by ctubbsii** [comment]
> In the for loop, the break condition is the same as the do-while loop meaning once the for loop is finished the do-while will only execute once(essentially rendering it useless).

It does call `waitForBalance` before entering this second loop, so even though the loop condition is the same, I think the expectation is that `waitForBalance` call could make the condition not true temporarily, hence the second loop. I'm not sure if this 5 seconds is enough time to guarantee that balance has actually started.

> After this happens, 2 servers are killed which sometimes takes a bit of time. This may be the window of time in which the tablets migrate under the radar.

Hopefully not. Once balancing has settled, there should be little reason for them to migrate.

> I have tried increasing this amount of time and haven't gotten the error since, however this error doesn't happen too often as is so its hard to tell if this is definitely the cause.

You could try shortening the time to see if the error appears more often afterwards. That might help indicate that increasing the time is definitely going to improve the stability of the test and is the correct path forward.

**Comment by milleruntime** [comment]
@DomGarguilo I agree with your assessments of that IT and I think you are on the right path.  Your analysis led me to look at what the Master is doing in `waitForBalance` and I noticed a TODO in the findbugs warning suppression: 
https://github.com/apache/accumulo/blob/bbd87a6693fc5cdbbe947b0821a4f05a18cf905b/server/manager/src/main/java/org/apache/accumulo/master/Master.java#L1594-L1608
It may be worthwhile to take a look at this code.

**Comment by DomGarguilo** [comment]
>     * the suspend time elapsed and the tablets are free to migrate before we are able to check (possibly because the tserver took too long to restart or WAL recovery on tablet load is causing the suspend time to elapse before we check the tablet states)
It seems like the suspend time elapsing could very well be the cause of this issue. I don't think that the time taken for the tserver to restart has bearing on this issue because the inconsistent tablet locations are recorded before tservers are restarted.
>     * the tserver fully recovered and then participated in subsequent migrations
I do not think this can be the case because the tservers are not restarted until after the issue-causing assert occurs. This error happens between the time the pre-death tablet location are noted and the point at which the suspended tablets are gathered, and then their tablets are compared to their pre-suspend location.
>     * we're splitting tablets and creating new migrations (possibly because splits weren't stabilized before killing the tserver)
I'm not sure that this can be the case because the splits happen near the start of the test, then time is allowed for migrations and such to finish. A printout of the tablet locations after balancing consistently shows the 3 servers used in this test with 10 tablets on each which seems balanced.
>     * the balancer is misbehaving, and rebalancing when it isn't supposed to
Seems very likely but I am having trouble looking into this.
>     * they migrated before we killed the tserver, but we didn't record it correctly in the test
This could be the case but does not seem likely. There is a brief window of time between when the tablet locations are recorded and the tservers are killed however I have tried allowing time before this happens for migrations to finish and it does not seem to increase the reliability of this test.

Hopefully I did not misinterpret any of your points, @ctubbsii. I'm also not sure why the formatting is messed up with quoting the bullets.

**Comment by ctubbsii** [comment]
Hmm, based on where it's failing, it seems like it's possible that not all tablets are yet marked as suspended. The loop condition on line 213 only waits until there's *some* suspended tablets for at least two tservers that were killed. However, it doesn't loop until *all* tablets hosted on those tservers are marked as suspended, which is the check on line 225 that is failing.

The comment above line 225 says "All suspended tablets should "belong" to the dead tablet servers". However, it's not actually checking for that... what it's checking for is that "all tablets belonging to the dead tablet servers are now suspended"... but that's not a condition that we waited for in the loop at line 213.

We can probably wait for all tablets assigned to the dead tserver to be marked as suspended OR we can relax the check on line 225 to only verify what is in the comment, and not assume that all of them that had been assigned to that tserver have been suspended yet.

This is just a hunch, though, that not all the tablets have yet been marked as suspended. You could verify this hunch by checking the state of the other tablets that were assigned to the dead tserver when the equality check on line 225 would have failed (something like replacing `assertEquals` with `if not equals, then dump ds.hosted and fail()`).

**Comment by ctubbsii** [comment]
It was suggested on a private call that extending the suspend duration might be helpful, because it may not be long enough. Also, it may help to ensure the system is initially balanced by creating the table with initial split points, rather than adding split points after the table is created. That could help avoid active migrations before the tservers are taken offline.
